README
# UIMA Text Segmenter
# Copyright (C) 2011  Nicolas Hernandez
#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
# 
#   http://www.apache.org/licenses/LICENSE-2.0
# 
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.  
#
# Fichier README
# 2011-02-28  Nicolas Hernandez  <nicolas.hernandez@gmail.com>


OVERVIEW

	For now, *UIMA Text Segmenter* is a [http://uima.apache.org/ UIMA] *wrapper 
	for the java implementations of the segmentation algorithms C99 and 
	TextTiling, written by Freddy Choi*, and used in the experiments described 
	in _choi_2000naacl_ (see below). These algorithms offer a text segmentation analysis
	at the discourse level essentially based on lexical cohesion measures.
	
	The UIMA wrapper uses the `uk.ac.man.cs.choif` package which comes from the 
	`C99-1.2-release.tgz` file, initially released by Freddy Choi.
	This package was released under the following license:
	_Free for educational, research and other non-profit making uses only. 
	Copyright 1999._
	All the source code is included in the current uima-text-segmenter release
	except the implementations of the Jeff Reynar's dotplotting algorithm 
	(reynar_1994, reynar_1998, reynar_1999) and Min-Yen Kan et al.'s Segmenter 
	(kan_et_al_1998) which were removed.
		
	License attached to the `fr.univnantes.lina.uima.textSegmenter` package
	is _Apache-2.0_.

	Consult the web page http://code.google.com/p/uima-text-segmenter for more information.
  
	{{{
	@InProceedings{choi_2000naacl,
  		author =       {Freddy Y. Y. Choi},
  		title =        {Advances in domain independent linear text segmentation},
  		booktitle =    {Proceedings of NAACL'00},
  		year =         {2000},
  		address =      {Seattle, USA},
  		month =        {April},
  		OPTannote =    {"choi00.ps.gz"}
	}
	}}}  
	
FEATURES


INSTALL & RUN

Si vous l'utilisez en tant que jar, alors 
  * créer un descripteur qui étende `desc/textSegmenter/C99AE` ou 
  `desc/textSegmenter/JTextTileAE` suivant que vous souhaitez utiliser C99 ou 
  JTextTile. 
  * Reportez vous à la section `parameters` ci-dessous pour connaître les 
  paramètres à étendre puis à définir.

Si vous récupérez les sources comme un projet Eclipse
Alors ajouter dans le build path
  * desc
  * resources
  * UIMA_HOME/addons/annotator/SnowballAnnotator/lib/uima-an-snowball.jar
  * UIMA_HOME/addons/annotator/WhitespaceTokenizer/lib/uima-an-wst.jar

Vous pouvez utiliser ensuite le documentAnalyser sur votre descripteur précédemment créé ou bien partir du descripteur exemple (voir Section EXAMPLE).


INSTALLATION PROBLEMS

Si vous avez un problème d'Access Restriction sur la classe sun.misc ou une de ses méthodes, 
Alors Properties > Java build path > Libraries > Remove JRE System Library > Add Library > JRE System Library 
(vous pouvez essayer de recharger le même par exemple le default) ; perso cela marche avec le jdk.1.6.17

Le code code source de `uk.ac.man.cs.choif.nlp.seg.linear.JTextTile` compile avec 
une erreur. Cette classe étant réécrite (voir section Description) ce n'est pas 
génant pour le développement ou l'utilisation du wrapper. Néanmoins pour 
éviter le signal de l'erreur, on a localement édité et  solutionne en 
initialisant à `null` la variable S dans la ligne `protected Stopword S = new Stopword();` 

Alors que le jar distribué sur le site contient les fichiers binaires, le dépot 
svn ne les contient pas. Il se peut que vous ayez des erreurs de compilation 
car tous les binaires n'ont pas été régénérés à partir de vos sources. 
Commencer d'abord par stopper l'auto build, puis à effectuer un jcasgen sur 
desc/textSegmenter/TextSegmenterTS.xml, puis un remove/add desc as build source 
et a la mano build project. A l'aide du navigator, vérifiez vous devez avoir 
dans le bin les répertoires suivants : fr (avec un sous répertoire types), uk, 

	
	
DESCRIPTION

	Les descripteurs des AE sont `desc/textSegmenter/C99AE` et `desc/textSegmenter/JTextTileAE`. 
	Ils fonctionnent respectivement avec les classes `fr.univnantes.lina.uima.textSegmenter.C99.C99AE` 
	et `fr.univnantes.lina.uima.textSegmenter.JTextTile.JTextTileAE`.
	 
	`desc/textSegmenter/TextSegmenterTS` définit un type d'annotation par 
	défaut correspondant à un analyseur de segmentation de texte au niveau discursif.
	Néanmoins `C99AE` et `JTextTileAE` définissent leur propre type.

	L'encapsulation de la bibliothèque du Choi a nécessité de faire des 
	modifications dans la classe `uk.ac.man.cs.choif.nlp.seg.linear.C99` 
	et `uk.ac.man.cs.choif.nlp.seg.linear.JTextTiling` de Choi. Certaines 
	méthodes privées ont été mises publiques et certaines finales ont été 
	autorisées à être redéfinies dans les sous-classes. Les méthodes main ont 
	été commentées ainsi que parfois certains constructeurs. 
	Les deux nouvelles classes sont respectivement `fr.univnantes.lina.uima.textSegmenter.C99.C99LINA` 
	et `fr.univnantes.lina.uima.textSegmenter.JTextTile.JTextTileLINA`.
	
	Nous notons que le code de Choi utilise l'API propriétaire de SUN en 
	particulier sun.misc.Sort.quicksort. Ce code est disponible dans la rt.jar (java runtimes 
	(Dans la 1.6 jre/lib/rt.jar)) avec un access restriction). Cette API ne devrait pas être utilisé. 
	JDK l'utilise pour des besoins internes néanmoins. Pour le supprimer... ici une piste 
	http://etutorials.org/Programming/Java+performance+tuning/Chapter+9.+Sorting/9.1+Avoiding+Unnecessary+Sorting+Overhead/

	La liste des mots vides pour le français est issue de http://snowball.tartarus.org
	wget http://snowball.tartarus.org/algorithms/french/stop.txt ; cat stop.txt|iconv -f iso-8859-1 -t utf8 | perl -ne 'chomp;/^([^ ]+)/; print "$1\n";' |grep -v '^$' > motsVides-fr.csv

LICENCES	
	
	See the Licenses* files
	
	
PARAMETERS

	Ci-dessous les paramètres communs et spécifiques que C99 et JTextile utilisent. 

	*Input|Output parameters* (Common to `desc/textSegmenter/C99AE` and `desc/textSegmenter/JTextTileAE`)
		* `InputView` CAS view to proceed. By default this view will be the `_InitialView`. If you do not now, then leaving it empty will probably be a good choice.
		* `InputSentenceAnnotation` Annotation type which corresponds to a sentence unit. By default `org.apache.uima.SentenceAnnotation`
		* `InputWordAnnotation`	Annotation type which corresponds to a word unit. By default will be `org.apache.uima.TokenAnnotation`
		* `InputWordFeature` Corresponds to the word token feature to consider. Default is `coveredText` but can be set to `stem` depending on the prior AE.
		* `OutputSegmentAnnotation` Annotation to create for each text segment detected. By default it is set to  `fr.univnantes.lina.uima.textSegmenter.types.SegmentAnnotation`. C99AE provides also a C99SegmentAnnotation while JTextTileAE provides a JTextTileAnnotation.
		* `StopWordFile` path to a stop word file
		
	*Algorithm parameters specific to `desc/textSegmenter/C99AE`*
		* `NumberOfSegments` 	Number of segments to make, if -1 then let the algorithm decide.		
		* `RankingMaskSize`	Size of ranking mask rank to smooth the similarity matrix of all the sentences (Apply hard ranking (replace pixel value with the proportion 
 of neighbouring values its greater than) to matrix using a S x S mask). It is said >=1 in the main method description and >=3 in the segment* method description. Surely should be >=3 to be effective. 
		* `TermFrequencyWeight`		Compute sentence similarity taking into account of term frequency (actually the entropy based on the token occurrence probability over all the sentences)
		
	*Algorithm parameters specific to `desc/textSegmenter/JTextTileAE`*
		* `WindowSize` Window size in number of words. Hearst suggested 30. Choi claimed  the same but its default value was 100 in its implementation. It depends on the number of words per sentence/proposition and also if we words with raw words, post tag filtered words, stems... and from the text genre...
Set it to -1 if you want the program to determine automatically the "best" value. The computation is based on  simple homemade empirical heuristic which set the best value to 3 times the average size of a sentence in terms of words.
		* `StepSize Size` of the step to shift the window. Hearst suggested 5. Choi claimed the same but its default value was 10 in its implementation. It depends on the number of words per sentence/proposition and also if we words with raw words, post tag filtered words, stems... and from the text genre...
Set it to -1 if you want the program to determine automatically the "best" value. The computation is based on  simple homemade empirical heuristic which set the best value to 1 time the average size of a sentence in terms of words. 


EXAMPLES

	`desc/textSegmenter/wst-snowball-C99-JTextTilingAAE.xml` constitue un exemple d'utilisation des descripteurs C99AE et JTextileAE.
	Il s'agit d'un aggregate qui utilise Apache WhitespaceTokenizer et Apache SnowballAnnotator pour fournir les annotations de phrases, mots et stem de mots.
	Cet exemple exécute à la fois C99 et JTextTiling l'un à la suite de l'autre. En pratique ces deux analyseurs n'ont aucune dépendance.

	Cet exemple requière l'ajout au classpath de  
	  * `UIMA_HOME/addons/annotator/SnowballAnnotator/lib/uima-an-snowball.jar`
      * `UIMA_HOME/addons/annotator/WhitespaceTokenizer/lib/uima-an-wst.jar`
 
	
TRICKS

MORE INFORMATION ABOUT WHAT IT IS WRAPPED

    The C99 algorithm works as following:
	Given a document as a list of elementary text blocks (usually tokenised 
	sentences), segment the document into NumberOfSegments coherent topic 
	segments. If NumberOfSegments param is -1, the algorithm will decide the 
	appropriate number of segments by monitoring the rate of increase in 
	segment density.
	The main steps are:
	1. Normalizes the sentences
	2. Computes a Similarity matrix of all the sentences
	3. Ranks the matrix (smoothing of each pixel by applying hard ranking 
	(replace pixel value with the proportion of neighbouring values its greater 
	than) to matrix using a S x S mask
	4. Sums of rank matrix
	5. Divides the matrix diagonal clustering which corresponds to determining 
	the	boundaries,	the number of boundaries depends on the user parameter (it 
	can	be done automatically)
	
	The TextTiling algorithm works as following:
	`TODO`

	
CONTACT

	See the header of this file.


ChangeLog
# UIMA Text Segmenter
# Copyright (C) 2011  Nicolas Hernandez
#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
# 
#   http://www.apache.org/licenses/LICENSE-2.0
# 
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.  
#
# Fichier ChangeLog
# 2011-02-28  Nicolas Hernandez  <nicolas.hernandez@gmail.com>

2011-06-17  Nicolas Hernandez  <nicolas.hernandez@gmail.com> committed
  * resources/stopWords-emptyList.csv: added 

2011-06-17  Nicolas Hernandez  <nicolas.hernandez@gmail.com> committed
  * add INSTALL file and refactor the README file and modifications comming from uima-common + add some stopwords files
  
2011-04-29  Nicolas Hernandez  <nicolas.hernandez@gmail.com> not tagged
  * uima-common: répercussions de refactoring en ajoutant dans le même package JCasSofaView, Annotation, AnnotationCollection, Feature et DocumentAnnotation Utilities, et en renommant nombreuses méthodes afin de les rendre plus explicites sur leur rôles
  
2011-03-13 Nicolas Hernandez  <nicolas.hernandez@gmail.com> non inclu dans le jar en ligne
  * Mise à jour de certains fichiers après clean et régénération du bin
  * README: quelques conseils pour la régénération pour éviter des problèmes de compilation

  
2011-02-28 Nicolas Hernandez  <nicolas.hernandez@gmail.com>
  * Ajout des fichiers  ChangeLog, CREDITS, KNOWN_BUGS, LICENSE, TODO et mise à jour du README
  * Ajour des headers licences aux fichiers src
  * Refactoring des noms des fichiers desc avec les suffixes AE, TS et AAE 
  * C99Final en C99 aussi, et C99WithSentenceWSTSnowball en C99, voir aussi les classes des src notamment ParserSentenceWSTSnowBallC99 en ParserC99
  * C99AE: paramétrage des types sentenceAnnotation et tokenAnnotation de C99 (suppression des AE et des types Paragraph et Sentence Segmenter)
  * ParserSentenceWSTSnowBallC99: paramétrage pour que l'on puisse spécifier le token feature sur lequel travailler (stem ou coveredText ou autre)
  * Définition du TS par défaut, SegmentAnnotation, au sein du desc C99AE à la place de celui en dur type fr.univnantes.lina.uima.jc99.ThemeSegment
  * Utilisation de uima-common pour faire étendre l'AE de AnnotatorEngine, permet de récupérer les inputView et outputAnnotation 
  * Nommage dans le descripteur des param OutputAnnotation en OutputSegmentAnnotation et Sentence/Token en InputSentence/TokenAnnotation puis créer un inputTokenFeatureAnnotation et retirer le outputFeature car If one of the parameter InputAnnotation or InputFeature (resp OutputAnnotation et OutputFeature) is defined, both must be ! at fr.univnantes.lina.uima.util.AnalysisEngine.initialize(AnalysisEngine.java:284)  
  * Paramétrisation des message debug à l'aide d'une propriétés pour C99AE et C99Parser
  * Ajout de commentaires dans le processInputView de C99AE et dans les en-têtes des méthodes de C99Parser
  * fichiers LICENSE*
  * ajout de la signification des parameters dans le README et dans la descritption des param dans le C99AE (d'après ce qui a été lu dans le main de la classe C99)
  * Déplacment de la partie de C99Parser qui prenait en charge le Parser du fait que sera aussi utilisé par JTextTile au sein de la classe UIMARawText qui étend RawText  
  * Ajout de la gestion de JTextTile
  * Externalisation des classes C99Lina et JTextTileLina qui sont des modifications des classes C99 et JTextTile du package de uk
  * Effacer C99Old.java et JTextTileOld.java de uk...seg.linear
   

# 2011-02-27 Nicolas Hernandez  <nicolas.hernandez@gmail.com>
Version de Christine Jacquin
  * n'utilise pas le SentenceAnnotation de wst mais celui de ParagraphSentenceAnnotator  
  * ParagraphSentenceA n'a dans son fonctionnement aucun lien avec wst (fonctinone avec des regex qui définissent Sentence et Paragraph), il produit les annotations Sentence et Paragraph, qui sont utilisées par C99WithSentenceWSTSnowball
  * ParagraphA n'est pas utilisé, 
  * D'après le descr on ne peut pas paraméter les types sur lesquels le C99WithSentenceWSTSnowball fonctionne
  * AE paragraph et sa classe fr.uima.ParagraphAnnotator ne sont pas utilisés
  * uk.ac.man.cs.choif.nlp.seg.linear.C99Lina est notre modification de C99, est utilisée par ParserSentenceWSTSnowBallC99, qui est utilisé par l'AE C99WithSentenceWSTSnowball
  * C99WithSentenceWSTSnowball utilisent a priori seulement les types Token et Sentence déclarés en final
  * Bug le segmenteur sentence ne reconnait pas toujours la dernière phrase d'un texte comme une phrase selon qu'elle termine directement un fichier.   

src/fr/univnantes/lina/uima/textSegmenter/JTextTile/JTextTileLINA.java
package fr.univnantes.lina.uima.textSegmenter.JTextTile;

import java.util.*;
import uk.ac.man.cs.choif.nlp.doc.basic.*;
import uk.ac.man.cs.choif.nlp.surface.*;
import uk.ac.man.cs.choif.extend.*;
/**
 * An implementation of Marti Hearst's text tiling algorithm.
 * Creation date: (07/12/99 01:31:36)
 * 
 * Classe modifiée par Nicolas Hernandez le 08/03/11
 * Principalement changement des signatures de certaines méthodes au niveau de 
 * la portée (les tâches principales sont exécutées par le main et le 
 * constructeur JTextTileLINA ; ce sont les méthodes appelées qui ont été 
 * rendues accessibles de l'extérieure de la classe), initialisation différente
 * de certaines variables, et mise en commentaire du main
 * 
 * classe originale: C99.java
 * @author: Freddy Choi
 */
public class JTextTileLINA {
	/* Program parameters */
	protected int w = 100; // Size of the sliding window
	protected int s = 10; // Step size
		
	/* Data sets */
	// Changed by N Hernandez: before initialization with constructor ( no Stopword() constructor with such signature  available), after to null value
	// before //protected Sentences C =  new RawText(); // A collection for segmentation
	protected Sentences C = null ;  
	protected Stopword S = null; // new Stopword(); // A stopword list for noise reduction
	

	/* Token -> stem dictionary */
	protected Hashtable stemOf = new Hashtable(); // Token -> stem

	/* Similarity scores and the corresponding locations */
	protected float[] sim_score = new float[0];
	protected int[] site_loc = new int[0];

	/* Depth scores */
	protected float[] depth_score = new float[0];

	/* Segment boundaries */
	public Vector boundaries = new Vector();
	
/**
 * Segment a collection
 * Creation date: (07/12/99 01:33:32)
 * @param c uk.ac.man.cs.choif.nlp.struct.RawText
 * @param s uk.ac.man.cs.choif.nlp.stopword.Stopword
 * @param w Window size
 * @param step Step size
 * 
 * Modified by Nicolas Hernandez 110308 :  commented
 */
	
/*	
public JTextTileLINA(Sentences c, Stopword s, int w, int step) {
	C = c;
	S = s;
	this.w = w;
	this.s = step;
	Debugx.msg("JTextTile", "Lemmatization...");
	preprocess();
	Debugx.msg("JTextTile", "Similarity scores...");
	similarityDetermination();	// Compute similarity scores
	Debugx.msg("JTextTile", "Depth scores...");
	depthScore();				// Compute depth scores using the similarity scores
	Debugx.msg("JTextTile", "Identifying boundaries...");
	boundaryIdentification();	// Identify the boundaries
	Debugx.msg("JTextTile", "Ready.");
}
*/
	
/**
 * Add a term to a block
 * Creation date: (07/12/99 01:41:24)
 * @param term java.lang.String
 * @param B java.util.Hashtable
 */
private static void blockAdd(final String term, Hashtable B) {
	Integer freq = (Integer) B.get(term);
	
	if (freq == null) freq = new Integer(1);
	else freq = new Integer(freq.intValue() + 1);

	B.put(term, freq);
}
/**
 * Compute the cosine similarity measure for two blocks
 * Creation date: (07/12/99 01:49:16)
 * @return float
 * @param B1 java.util.Hashtable
 * @param B2 java.util.Hashtable
 */
private static float blockCosine(final Hashtable B1, final Hashtable B2) {
	/* 1. Declare variables */
	int W; // Weight of a term (temporary variable)
	int sq_b1 = 0; // Sum of squared weights for B1
	int sq_b2 = 0; // Sum of squared weights for B2
	int sum_b = 0; // Sum of product of weights for common terms in B1 and B2
	
	/* 2. Compute the squared sum of term weights for B1 */
	for (Enumeration e=B1.elements(); e.hasMoreElements();) {
		W = ((Integer) e.nextElement()).intValue();
		sq_b1 += (W * W);
	}

	/* 3. Compute the squared sum of term weights for B2 */
	for (Enumeration e=B2.elements(); e.hasMoreElements();) {
		W = ((Integer) e.nextElement()).intValue();
		sq_b2 += (W * W);
	}

	/* 4. Compute sum of term weights for common terms in B1 and B2 */

	/* 4.1. Union of terms in B1 and B2 */
	Hashtable union = new Hashtable(B1.size() + B2.size());
	for (Enumeration e=B1.keys(); e.hasMoreElements();) union.put((String) e.nextElement(), new Boolean(true));
	for (Enumeration e=B2.keys(); e.hasMoreElements();) union.put((String) e.nextElement(), new Boolean(true));

	/* 4.2. Compute sum */
	Integer W1; // Weight of a term in B1 (temporary variable)
	Integer W2; // Weight of a term in B2 (temporary variable)
	String term; // A term (temporary variable)
	for (Enumeration e=union.keys(); e.hasMoreElements();) {
		term = (String) e.nextElement();
		W1 = (Integer) B1.get(term);
		W2 = (Integer) B2.get(term);
		if (W1!=null && W2!=null) sum_b += (W1.intValue() * W2.intValue());
	}
	
	/* 5. Compute similarity */
	float sim;
	sim = (float) sum_b / (float) Math.sqrt(sq_b1 * sq_b2);
		
	return sim;
}
/**
 * Remove a term from the block
 * Creation date: (07/12/99 01:46:39)
 * @param term java.lang.String
 * @param B java.util.Hashtable
 */
private static void blockRemove(final String term, Hashtable B) {
	Integer freq = (Integer) B.get(term);

	if (freq != null) {
		if (freq.intValue() == 1) B.remove(term);
		else B.put(term, new Integer(freq.intValue() - 1));
	}
}
/**
 * Identify the boundaries
 * Creation date: (07/12/99 07:05:04)
 * 
 * Modified by Nicolas Hernandez 110308 :  private -> protected
 */
protected void boundaryIdentification() {
	/* Declare variables */
	float mean = 0; // Mean depth score
	float sd = 0; // S.D. of depth score
	float threshold; // Threshold to use for determining boundaries
	int neighbours = 3; // The area to check before assigning boundary
	
	/* Compute mean and s.d. from depth scores */
	for (int i=depth_score.length; i-->0;) mean += depth_score[i];
	mean = mean / depth_score.length;

	for (int i=depth_score.length; i-->0;) sd += Math.pow(depth_score[i] - mean, 2);
	sd = sd / depth_score.length;

	/* Compute threshold */
	threshold = mean - sd / 2;

	/* Identify segments in pseudo-sentence terms */
	Vector pseudo_boundaries = new Vector();
	boolean largest = true; // Is the potential boundary the largest in the local area?
	for (int i=depth_score.length; i-->0;) {

		/* Found a potential boundary */
		if (depth_score[i] >= threshold) {
		
			/* Check if the nearby area has anything better */
			largest = true;
			
			/* Scan left */
			for (int j=neighbours; largest && j>0 && (i-j)>0; j--) {
				if (depth_score[i-j] > depth_score[i]) largest=false;
			}

			/* Scan right */
			for (int j=neighbours; largest && j>0 && (i+j)<depth_score.length; j--) {
				if (depth_score[i+j] > depth_score[i]) largest=false;
			}

			/* Lets make the decision */
			if (largest) pseudo_boundaries.addElement(new Integer(site_loc[i]));
		}
	}
	
	/* Convert pseudo boundaries into real boundaries.
	We use the nearest true boundary. */

	/* Convert real boundaries into array for faster access */
	int[] true_boundaries = new int[C.sentenceBoundaries().size()];
	for (int i=true_boundaries.length; i-->0;) true_boundaries[i]= ((Integer) C.sentenceBoundaries().elementAt(i)).intValue();
	
	int pseudo_boundary;
	int distance; // Distance between pseudo and true boundary
	int smallest_distance; // Shortest distance
	int closest_boundary; // Nearest real boundary
	for (int i=pseudo_boundaries.size(); i-->0;) {
		pseudo_boundary = ((Integer) pseudo_boundaries.elementAt(i)).intValue();

		/* This is pretty moronic, but it works. Can definitely be improved */
		smallest_distance = Integer.MAX_VALUE;
		closest_boundary = true_boundaries[0];
		for (int j=true_boundaries.length; j-->0;) {
			distance = Math.abs(true_boundaries[j] - pseudo_boundary);
			if (distance <= smallest_distance) {
				smallest_distance = distance;
				closest_boundary = true_boundaries[j];
			}
		}

		boundaries.addElement(new Integer(closest_boundary));
	}
}
/**
 * Compute depth score after applying similarityDetermination()
 * Creation date: (07/12/99 06:54:32)
 * 
 * Modified by Nicolas Hernandez 110308 :  private -> protected
 */
protected void depthScore() {
	/* Declare variables */
	float maxima = 0; // Local maxima
	float dleft = 0; // Difference for the left side
	float dright = 0; // Difference for the right side

	/* For each position, compute depth score */
	depth_score = new float[sim_score.length];
	for (int i=sim_score.length; i-->0;) {

		/* Scan left */
		maxima = sim_score[i];
		for (int j=i; j>0 && sim_score[j] >= maxima; j--) maxima = sim_score[j];
		dleft = maxima - sim_score[i];

		/* Scan right */
		maxima = sim_score[i];
		for (int j=i; j<sim_score.length && sim_score[j] >= maxima; j++) maxima = sim_score[j];
		dright = maxima - sim_score[i];

		/* Declare depth score */
		depth_score[i] = dleft + dright;
	}
}
/**
 * Generate text output with topic boundary markers.
 * Creation date: (07/12/99 07:39:00)
 */
private static void genOutput(Sentences c, Vector seg) {
	/* Declare variables */
	Vector text = c.tokens(); // The text
	Vector sentence = c.sentenceBoundaries(); // Sentence boundaries
	String line;
	int start, end; // Sentence boundaries
	
	/* The implicit boundary at the beginning of the file */
	System.out.println("==========");

	/* Print all the sentences */
	for (int i=1; i<sentence.size(); i++) {

		/* Get sentence boundaries */
		start = ((Integer) sentence.elementAt(i-1)).intValue();
		end = ((Integer) sentence.elementAt(i)).intValue();

		/* If start is a topic boundary, print marker */
		if (seg.contains(new Integer(start))) System.out.println("==========");

		/* Print a sentence */
		line = "";
		for (int j=start; j<end; j++) line += (text.elementAt(j) + " ");
		System.out.println(line.trim());
	}

	/* The implicit boundary at the end of the file */
	System.out.println("==========");
}
/**
 * Decide whether word i is worth using as feature for segmentation.
 * Creation date: (07/12/99 23:39:51)
 * @return boolean
 * @param i int
 */
private boolean include(int i) {
	/* Noise reduction by filtering out everything but nouns and verbs - 
	Best but requires POS tagging
	String pos = (String) C.pos.elementAt(i);
	return (pos.startsWith("N") || pos.startsWith("V")); */
	/* Noise reduction by stopword removal - OK */
	String token = (String) C.tokens().elementAt(i);
	return !S.isStopword(token.toLowerCase());

	/* No noise reduction -- Worst
	return true; */
}
/**
 * 
 * Creation date: (07/12/99 04:20:08)
 * @param args java.lang.String[]
 * 
 * Modified by Nicolas Hernandez 110308 :  commented (embedded comments adapted)
 */
/*
public static void main(String[] args) {
	// Print header 
	Debugx.header("This is JTextTile, a Java implementation of Marti Hearst's TextTiling algorithm.");

	Argx arg = new Argx(args);
	int window = arg.get("-w", 30, "Window size");
	int step = arg.get("-s", 5 , "Step size");
	String stopwordList = arg.get("-sw", Stopword.defaultResourceFile, "Path and name of stopword list");
	arg.displayHelp();
	
	try {
		Debugx.msg("Window", window);
		Debugx.msg("Step", step);
		Debugx.msg("Stopword list", stopwordList);

		// Load data 
		Stopword s = new Stopword(new java.io.File(stopwordList));
		RawText c = new RawText(System.in);

		// A bit of error checking 
		Debugx.msg("Collection", c.text.size());
		if (c.text.size() <= (window * 2)) {
			Debugx.msg("Fatal error", "Window size (" + window + " * 2 = " + (window * 2) + ") larger then collection (" + c.text.size() + ")");
			System.exit(1);
		}

		// Lets boogie 
		JTextTileLINA t = new JTextTileLINA(c, s, window, step);		// Initialise text tiling algorithm with collection
		JTextTileLINA.genOutput(c, t.boundaries);				// Generate segmented output
		System.exit(0);
	}
	catch (Exception e) {
		Debugx.handle(e);
		System.exit(1);
	}
}
*/


/**
 * Perform some preprocessing to save execution time
 * Creation date: (07/12/99 03:21:34)
 * 
 * Modified by Nicolas Hernandez 110308 :  private -> protected
 */
protected void preprocess() {
	/* Declare variables */
	Vector text = C.tokens(); // Text of the collection
	String token; // A token

	/* Construct a dictionary of tokens */
	for (int i=text.size(); i-->0;) {
		token = (String) text.elementAt(i);
		stemOf.put(token, new Integer(0));
	}

	/* Complete mapping token -> stem */
	for (Enumeration e=stemOf.keys(); e.hasMoreElements();) {
		token = (String) e.nextElement();
		stemOf.put(token, Stemmer.stemOf(token));
	}
}
/**
 * Compute the similarity score.
 * Creation date: (07/12/99 03:17:31)
 * 
 * Modified by Nicolas Hernandez 110308 :  private -> protected
 */
protected void similarityDetermination() {
	/* Declare variables */
	Vector text = C.tokens(); // The source text
	Hashtable left = new Hashtable(); // Left sliding window
	Hashtable right = new Hashtable(); // Right sliding window
	Vector score = new Vector(); // Scores
	Vector site = new Vector(); // Locations

	/* Initialise windows */
	for (int i=w; i-->0;) blockAdd((String) stemOf.get((String) text.elementAt(i)), left);
	for (int i=w*2; i-->w;) blockAdd((String) stemOf.get((String) text.elementAt(i)), right);

	/* Slide window and compute score */
	final int end = text.size() - w; // Last index to check
	String token; //  A stem
	int step=0; // Step counter
	int i; // Counter

	for (i=w; i<end; i++) {
		/* Compute score for a step */
		if (step == 0) {
			score.addElement(new Float(blockCosine(left, right)));
			site.addElement(new Integer(i));
			step = s;
		}

		/* Remove word which is at the very left of the left window */
		if (include(i-w)) {
			blockRemove((String) stemOf.get((String) text.elementAt(i-w)), left);
		}

		/* Add current word to the left window and remove it from the right window */
		if (include(i)) {
			token = (String) text.elementAt(i);
			blockAdd((String) stemOf.get(token), left);
			blockRemove((String) stemOf.get(token), right);
		}

		/* Add the first word after the very right of the right window */
		if (include(i+w)) {
			blockAdd((String) stemOf.get((String) text.elementAt(i+w)), right);
		}

		step--;
	}
	/* Compute score for the last step */
	if (step == 0) {
		score.addElement(new Float(blockCosine(left, right)));
		site.addElement(new Integer(i));
		step = s;
	}

	/* Smoothing with a window size of 3 */
	sim_score = new float[score.size()-2];
	site_loc = new int[site.size()-2];
	for (int j=0; j<sim_score.length; j++) {
		sim_score[j] = (((Float) score.elementAt(j)).floatValue() + ((Float) score.elementAt(j+1)).floatValue() + ((Float) score.elementAt(j+2)).floatValue()) / 3;
		site_loc[j] = ((Integer) site.elementAt(j+1)).intValue();
	}
	
}
}
